### PR TITLE
LC30CHALLENGE: Problem 15

### PR DESCRIPTION
**Pull request by DmitryBokovikov**

### CODE REVIEW

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem15.cpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#include "stdafx.h"
+#include "problems.h"
+
+namespace problems_ns {
+std::vector<int> productExceptSelf(std::vector<int>& nums) {
+  const int n = size(nums);
+  std::vector<int> ans(n);
+  std::partial_sum(cbegin(nums), cend(nums), begin(ans), std::multiplies<int>());
```
Comment (fbokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem15.cpp:8]: What a misleading name of function :(

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem16.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#include "stdafx.h"
+#include "problems.h"
+
+namespace problems_ns {
+bool checkValidString(std::string& s) {
```
Comment (fbokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem16.cpp:5]: Haven't understood at all your solution :(

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem17.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace problems_ns {
+int numIslands(std::vector<std::vector<char>>& grid) {
+  const int n = size(grid);
+  const int m = empty(grid) ? 0 : size(grid.front());
+
+  std::function<void(int, int)> dfs;
+  dfs = [&](int i, int j) {
+    grid[i][j] = 'X';
+    for (auto[x, y] : std::vector<std::pair<int, int>>{{i + 1, j}, {i - 1, j}, {i, j + 1}, {i, j - 1}})
```
Comment (fbokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem17.cpp:12]: A-ha-ha! Used exactly the same patterns

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem16.cpp (hunk 2 above)
Comment (DmitryBokovikov) [reply src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem16.cpp:5]: Added dp solution for your convenience

**Pull request merged**